fe/fe-core/src/main/java/com/starrocks/lake/StarMgrMetaSyncer.java

```diff
@@ -236,72 +237,139 @@ public int deleteUnusedWorker() {
-        Database db = GlobalStateMgr.getCurrentState().getDb(dbName);
-        if (db == null) {
-            throw new DdlException(String.format("db %s does not exist.", dbName));
+            List<Table> tables = db.getTables();
+            for (Table table : tables) {
+                if (!table.isCloudNativeTableOrMaterializedView()) {
+                    continue;
+                }
+                try {
+                    syncTableMetaAndColocationInfoInternal(db, (OlapTable) table, true /* forceDeleteData */);
+                } catch (Exception e) {
+                    LOG.info("fail to sync table {} meta, {}", table.getName(), e.getMessage());
+                }
+            }
         }
+    }
 
-        HashMap<Long, List<Long>> feGroupToShards = new HashMap<>();
+    // return true if starmgr shard meta changed
+    private boolean syncTableMetaInternal(Database db, OlapTable table, boolean forceDeleteData) throws DdlException {
+        StarOSAgent starOSAgent = GlobalStateMgr.getCurrentStarOSAgent();
+        HashMap<Long, Set<Long>> redundantGroupToShards = new HashMap<>();
+        Collection<PhysicalPartition> physicalPartitions = null;
         Locker locker = new Locker();
         locker.lockDatabase(db, LockType.READ);
         try {
-            Table table = db.getTable(tableName);
-            if (table == null) {
-                throw new DdlException(String.format("table %s does not exist.", tableName));
-            }
-            if (!table.isCloudNativeTableOrMaterializedView()) {
-                throw new DdlException("only support cloud table or cloud mv.");
+            if (db.getTable(table.getId()) == null) {
+                return false; // table might be dropped
             }
+            physicalPartitions = table.getAllPhysicalPartitions();
+        } finally {
+            locker.unLockDatabase(db, LockType.READ);
+        }
 
-            OlapTable olapTable = (OlapTable) table;
-            for (Partition partition : olapTable.getAllPartitions()) {
-                for (PhysicalPartition physicalPartition : partition.getSubPartitions()) {
-                    long groupId = physicalPartition.getShardGroupId();
-                    List<Long> feShardIds = new ArrayList<>();
-                    for (MaterializedIndex materializedIndex :
-                            physicalPartition.getMaterializedIndices(MaterializedIndex.IndexExtState.ALL)) {
-                        for (Tablet tablet : materializedIndex.getTablets()) {
-                            feShardIds.add(tablet.getId());
-                        }
-                    }
-                    if (!feShardIds.isEmpty()) {
-                        feGroupToShards.put(groupId, feShardIds);
+        for (PhysicalPartition physicalPartition : physicalPartitions) {
+            locker.lockDatabase(db, LockType.READ);
+            try {
+                if (table.getState() != OlapTable.OlapTableState.NORMAL) {
```

Comment:
whether need to check the db and table exists again, since this is a new area of the lock?

Reply:
not necessary, even if db/table/partition dropped, this still work, too many check might cause confusion.